Enforce server root dir existence with an optional opt-out arg

black-dragon74 opened the pull request:
By default the program will now verify if the server root directory really exists or not. It can help nullify the cases where the config file simply won't work because of a typo in the root dir path.

However there might be cases where the user is trying to create config file to use on a separate machine in such cases the validation can easily be skipped by using `-s` or `-skiproot` or `--skiproot` args.

This commit also refactors some code and enforces proper exit methods instead of using the improper `return` keyword.

black-dragon74 committed:
Enforce server root dir existence with an optional opt-out arg

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

Sid-Sun commented:
Hey, nice feature!

black-dragon74 commented:
Do not merge it yet. I have a better implementation. Will push that commit shortly.

Sid-Sun commented:
However, I notice version number is being updated to 3.2 but this is a change to the application's functionality and therefore is a major upgrade (i.e. it should now be 4.0)

Secondly, I notice avoidable replications of ox.Exit in the argument check block, we should add it outside of the specific argument scopes (i.e. the outermost IF block) (the same is true for the getConsent IF statement towards the end of the function)

Speaking of os.Exit, we try to avoid that in our case because unlike returns, os.Exit does not wait for defers to finish (while we don't use them here currently, in case we do it may resolve to undesired behavior 

If the selection is 9 (i.e the user wants to exit, we should not exit with 1, since no error happened)

Also, if the dir path doesn't exist, there should be an iteration that asks the user to re-enter the root path they desire (reducing the hassle of re-entering)

Additionally, the variable name `pArgs` only has one argument therefore is not very proper 

I forgot to update help to show the availability of printing version, if you could add it for me, it'd be great :)

Looking forward to the changes (and the better implementation I guess :P )

Sid-Sun commented:
I suppose I'll start making PRs too, make sure I don't forget to update help descriptions :P

Sid-Sun reviewed (changes requested):
I notice version number is being updated to 3.2 but this is a change to the application's functionality and therefore is a major upgrade (i.e. it should now be 4.0)

Secondly, I notice avoidable replications of ox.Exit in the argument check block, we should add it outside of the specific argument scopes (i.e. the outermost IF block) (the same is true for the getConsent IF statement towards the end of the function)

Speaking of os.Exit, we try to avoid that in our case because unlike returns, os.Exit does not wait for defers to finish (while we don't use them here currently, in case we do it may resolve to undesired behavior

If the selection is 9 (i.e the user wants to exit, we should not exit with 1, since no error happened)

Also, if the dir path doesn't exist, there should be an iteration that asks the user to re-enter the root path they desire (reducing the hassle of re-entering)

Additionally, the variable name pArgs only has one argument therefore is not very proper

black-dragon74 committed:
Several updates and enhancements related to server root dir validation

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

black-dragon74 commented:
1. Noted and fixed. Now I know more about how this program is versioned. Will take care in future PRs.

2. The repetitions were intentional as we wanted program to continue when a `-s` arg was provided but the way on which program was originally written the args were supposed to terminate program after a specific input, not continue it. However, the approach has now been re-written interactively.

3. The fact is known and the use of `os.Exit()` is encouraged as the program does not use `defer` as of now (like you agreed). The use of `runtime.Goexit()` would be overkill as we do not have multiple `goroutines`. So, let's keep them in place for now.

4. Fixed, totally missed it earlier.

5. The original idea was the same but we were asking for inputs serially and hence recursion needed a different approach. In commit https://github.com/Sid-Sun/nginx-auto-config/pull/2/commits/cd4e99abf9decac2bdbe318a17c6bfc36fa39265 I took the shorter route. The implementation has now been re-written to satisfy the original idea.

**Some additional improvements include:**
- Refactored `getConsent()` function for a much cleaner code.
- Added support for getting version info by using the standard `-v` arg. The non standard `version` arg has been removed.
- Updated workflows to reflect the changes in getting the program version 
- A constant `sVers` is declared at the top of the file to store program version number and hence avoiding the need to dig deeper into code.
- `fmt.Println()` is used instead of using `fmt.Printf()` with a `\n` delimiter to print the program version.

Sid-Sun commented on nginx-auto-config.go:
```diff
@@ -30,22 +32,22 @@ var red = color.New(color.FgRed)
 		} else {
-			fmt.Printf("Unknown option(s) %s, run with -h, -help or --help to get help, -s, -skiproot or --skiproot to skip server root directory validation or without any argumets to launch the program\n", pArgs)
+			fmt.Printf(
+					"Unknown option(s) %s\n" +
```
option**, only one argument is taken in the `pArg` var

black-dragon74 replied:
Rather cosmetic change. Better to take note of it and fix it with some other significant commit.

We are saying "option(s)" not "options" so it will do fine.

Sid-Sun commented on nginx-auto-config.go:
```diff
@@ -30,22 +32,22 @@ var red = color.New(color.FgRed)
+			fmt.Printf(
+					"Unknown option(s) %s\n" +
+						"Run with -h, -help or --help to get help\n" +
+						"-v or version to get program version\n" +
```
"version" won't work as per line #40

black-dragon74 replied:
Fixed

Sid-Sun commented on nginx-auto-config.go:
```diff
@@ -225,3 +229,15 @@ func takeInput() int {
+	if !dirExists(dirName) {
+		_, _ = red.Printf("Directory '%v' is non existent, please try again.\n", dirName)
+		_, _ = cyan.Print("Root path: ")
+		return verifyDirInput()
```
It'd be better to do:
```
if dirExists(dirName) {
    return dirName
}
_, _ = red.Printf("Directory '%v' is non existent, please try again.\n", dirName)
_, _ = cyan.Print("Root path: ")
return verifyDirInput()
```

black-dragon74 replied:
Underlying functionality is exactly the same and hence would be unnecessary.

Sid-Sun replied:
We should avoid using else as much as possible, it's not very clean

Sid-Sun replied:
You may notice this is a general theme followed in the code

black-dragon74 replied:
Done.

Sid-Sun commented on nginx-auto-config.go:
```diff
@@ -225,3 +229,15 @@ func takeInput() int {
+	} else {
+		return dirName
+	}
+}
```
I'd suggest moving this to utils.go

black-dragon74 replied:
I initially thought of the same but the `takeInput` was defined in the main file. Hence this function too was defined in the same file. However, moved.

Sid-Sun commented on utils.go:
```diff
@@ -37,10 +37,7 @@ func getInput(EmptyAllowed bool, SingleWorded bool) string {
-		return true
-	}
-	return false
+	return strings.ToLower(consent) == "y"
```
Nice catch!

black-dragon74 committed:
Cosmetic changes and basic re-structuring of code

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

black-dragon74 commented:
All the required changes have been made.

black-dragon74 committed:
Author hates else keyword

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

Sid-Sun commented:
Good contrib, will do a final review and if all's good, will merge

Sid-Sun commented on nginx-auto-config.go:
```diff
@@ -95,7 +114,10 @@ func main() {
 		if server.port == 443 {
 			printCautionSSL()
 		}
+		os.Exit(0)
```
This is not needed, there's an exit immediately after the block

black-dragon74 replied:
Done

black-dragon74 committed:
Removed unnecessary exit

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

Sid-Sun commented:
Changes Merged and deployed in version 4.0
